### PR TITLE
Online expand improvement

### PR DESCRIPTION
**Pull request by pengzhout**
Online expand improvement

    commit 8eed421738 & e0b06678aa allow us to do a gpexpand of GPDB cluster
    without a restart, so we called this strategy as "online expand", those
    two commits were mainly focus on how to avoid restarting cluster when
    expanding cluster, this commit will do the left work to improve gpexpand
    using a few features we merged to master recently.

    First improvement is, it's no longer necessary to change policy of all
    non-partition table to random when doing expanding. Previously, we couldn't
    tell the difference between expanded table and non-expaned table, so if
    the distribution policy of these tables were the same, planner took these
    tables's data as co-located and produced a incorrect plan. To avoid this,
    gpexpand used to change policy of all table to random so planner could
    produce a correct but a non-effective plan, because random policy always
    means annoying broadcase motion.

    Now each table has a 'numsegments' attribute introduced by 4eb65a53, GPDB
    can recognize expanded and non-expand table and produced correct plans. so
    the first imporvement is removing the randomization step of tables.

    The second improvement is, use a brand new syntax "ALTER TABLE foo EXPAND TABLE"
    to focus on the data rebalance of tables. Previously, tables were converted
    to randomly before rebalance and numsegments of tables were always the same
    with GPDB cluster size, so we can use a tricky "ALTER TABLE foo SET WITH
    (REORGNIZE = true) DISTRIBUTED BY (original_key)" command to rebalance data
    to new added segments, now policy and numsegments are not changed before
    rebalancing data, so expanding such tables is not match the concept of "SET
    DISTRIBUTED BY" now, we need a proper syntax to focus on table expanding.

    New expand syntax named "ALTER TABLE foo EXPAND TABLE", we have two methods
    to do the real data movement inside, one is CTAS, another is RESHUFFLE, which
    one is better depends on how much data needs to be moved and whether the table
    has index and whether the table is a append only table (analyzed by Hekki).

    The third improvement is, postpone the randomization of partition tables.
    Partition table stores data on leaf partitions only, previously, to expanded
    leaf partitions in parallel, gpexpand randomize policy of leaf part only and
    keep the policy of root/interior part no changed, SELECT/UPDATE/INSERT
    operations on the partition table work well in this mode, even some leaf
    partition was expaned and other leaf partitons were not.

    We want to keep the mode to expand partition table in parallel but postpone
    the randomization so planner can still use the policy info to produce
    effective plan before partition table is actually expanded. Another thing
    is we are online expand now, using utility mode directly is not safe, we need
    to add lock protection and keep the consistency of gp_distribution_policy
    when doing the randomization. As a result, "ALTER TABLE root_p EXPAND PREPARE"
    is designed to do the randomization and optimizations are added to let
    EXPAND PREPARE only be called one time for a partition table.

### CODE REVIEW

**Comment by pengzhout** [comment]
@hlinnaka Please help take a look, I guess you are interest in this expand topic and the methods to move data. hope I have got your point in #6260

**Comment by hlinnaka** [comment]
* I like the refactoring in `tablecmds.c`
* I like changing the `ALTER TABLE WITH (reshuffle)` syntax to `ALTER TABLE EXPAND`.
* I still don't understand what the EXPAND PREPARE command does, or why it's needed
* I don't understand what the problem and idea with partitions is.

Can you split this PR into two patches:
1. Refactor tablecmds.c, and change syntax to ALTER TABLE EXPAND.
2. Add ALTER TABLE EXTEND PREPARE command.

I think we could push the 1st part fairly quickly, and we could continue the discussion on ALTER TABLE EXPAND PREPARE and how partitions are handled.

**Comment by kainwen** [comment]
> * I like the refactoring in `tablecmds.c`
> * I like changing the `ALTER TABLE WITH (reshuffle)` syntax to `ALTER TABLE EXPAND`.
> * I still don't understand what the EXPAND PREPARE command does, or why it's needed
> * I don't understand what the problem and idea with partitions is.

Hi @hlinnaka , `set distributed by` will hold AccessExclusive Lock on the table. It is ideal to do it one leaf table by one, each in a single transaction so that users can still access other partitions(including insert, update, delete).

For hash-distributed  partition table, we should first:
1. increase all parts(root, middle-level, leaft) tables' numsegment
2. change leaf partitions to randomly distributed
These two steps are just what expand prepared does in a single transaction.

Then we can expand leaf table one by one. And during this process, access to other leaf partition is allowed(all CURD).

All the work is to avoid a single big transaction to expand all the leaf partitions and its root.

> 
> Can you split this PR into two patches:
> 
> 1. Refactor tablecmds.c, and change syntax to ALTER TABLE EXPAND.
> 2. Add ALTER TABLE EXTEND PREPARE command.
> 
> I think we could push the 1st part fairly quickly, and we could continue the discussion on ALTER TABLE EXPAND PREPARE and how partitions are handled.

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -14547,6 +14623,414 @@ ReshuffleRelationData(Relation rel)
...
+	 * record in ao visibility map table
+	 */
+	if (relstorage_is_ao(rel->rd_rel->relstorage))
+		return EXPANDMETHOD_CTAS;
+
+	/*
+	 * for partition table, use CTAS for now, EXPANDMETHOD_UPDATE
+	 * cannot expand randomly distributed partition table well
+	 */
+	if (ps != PART_STATUS_NONE)
```
[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/commands/tablecmds.c]: Indeed, there is a tiny assertion failure when using UPDATE method to expand random partition table, should not be a problem, I will check the assertion with team tomorrow, the restriction can be removed soon.

[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/commands/tablecmds.c]: Same here, I want to settle down the percentage of data movement that we begin to choose UPDATE method, any suggestion? in case we set it to a unreasonable value, thanks.

**Comment by hlinnaka** [comment]
On 21/11/2018 16:36, Zhenghua Lyu wrote:
>       * I like the refactoring in |tablecmds.c|
>       * I like changing the |ALTER TABLE WITH (reshuffle)| syntax to
>         |ALTER TABLE EXPAND|.
>       * I still don't understand what the EXPAND PREPARE command does,
>         or why it's needed
>       * I don't understand what the problem and idea with partitions is.
> 
> Hi @hlinnaka <https://github.com/hlinnaka> , |set distributed by| will 
> hold AccessExclusive Lock on the table. It is ideal to do it one leaf 
> table by one, each in a single transaction so that users can still 
> access other partitions(including insert, update, delete).

That's an optimization. Let's make it work, first. Minimum Viable 
Product and all that..

> For hash-distributed partition table, we should first:
> 
>  1. increase all parts(root, middle-level, leaft) tables' numsegment
>  2. change leaf partitions to randomly distributed
>     These two steps are just what expand prepared does in a single
>     transaction.
> 
> Then we can expand leaf table one by one. And during this process, 
> access to other leaf partition is allowed(all CURD).
> 
> All the work is to avoid a single big transaction to expand all the leaf 
> partitions and its root.

You can do all that as part of the ALTER TABLE EXPAND command, if you 
wish. I don't think we need to expose special syntax to the users for it.

But wouldn't it be simpler to just ALTER TABLE EXPAND each partition 
separately? Even in your proposal, you have an intermediate state where 
there's a mismatch in the distribution policy between the root partition 
and some leaves.

- Heikki

**Comment by pengzhout** [comment]
> You can do all that as part of the ALTER TABLE EXPAND command, if you wish. I don't think we need to expose special syntax to the users for it. But wouldn't it be simpler to just ALTER TABLE EXPAND each partition separately? Even in your proposal, you have an intermediate state where there's a mismatch in the distribution policy between the root partition and some leaves. - Heikki

Ahh, that's the best part of how old gpexand do the expand of partition table and it's the part we want to inherit from old gpexpand.  It's delicately designed, so please give me a try to explain it.

Let's say there is a moment after cluster is expaned from 2 to 3 and root_part is prepared, 

root_part (numsegments = 3, policy = hash)
  |__ leaf_part_1  (numsegments = 3, policy = random)
  |__ leaf_part_2  (numsegments = 3, policy = hash)

leaf_part_2 has been expanded, leaf_part_1 hasn't yet.  consider SELECT/UPDATE/INSERT:
SELECT:  select * from root_part.   APPEND node of scan of leaf_part_2 and leaf_part1 has a STRAW locus because the policy of subpaths is different, so plan should be correct.  
UPDATE: same as select.
INSERT:  plan is created according to the policy of root_part, leaf_part_1 don't care the data got, leaf_part_2 is expanded, so it has the same policy with root partition, so it's data is consistent with it's policy. 

To sum up, the mode allow a partition been expanded one by one and can still do INSERT/UPDATE/SELECT in middle stage correctly. That's the best part, if we didn't do a prepare work on root
partition, we cannot get the benefit mentioned and even worse the plan is incorrect in the middle stage.

Eg:
root_part (numsegments = 2, policy = hash)
  |__ leaf_part_1  (numsegments = 2, policy = hash)
  |__ leaf_part_2  (numsegments = 3, policy = hash)

In such moment, insert into root_part will make data and policy of leaf_part_2 inconsistent.

**Comment by pengzhout** [comment]
> You can do all that as part of the ALTER TABLE EXPAND command, if you wish.

Logically, we can do it inside ALTER TABLE EXPAND, my concern is those operations need acquiring an Exclusive lock on root part and other sibling leaf part, so it might block other expand job of sibling leaf part. gpexpand can do the expand job in parallel, so gpexpand did the prepare work one time and then can do leaf part expand in parallel.

File: src/backend/commands/tablecmds.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielgustafsson) [reply src/backend/commands/tablecmds.c]: If there is a known bug in the code, isn't it better to fix that before committing it?

File: src/backend/commands/tablecmds.c [hunk 2]
```diff
@@ -14547,6 +14623,414 @@ ReshuffleRelationData(Relation rel)
...
+	if (!prepare)
+	{
+		/* two ways to move data */
+		if (method == EXPANDMETHOD_CTAS)
+			ATExecExpandTableCTAS(rootCmd, rel, cmd, ps);	
+		else if (method == EXPANDMETHOD_UPDATE)
+		{
+			/* tricky way to make it same within ATExecExpandTableCTAS() */
+			ATExecExpandTableReshuffle(rootCmd, rel, cmd, ps);
+			relation_close(rel, NoLock);
```
Comment (danielgustafsson) [inline src/backend/commands/tablecmds.c]: Why is the relation open coming back here?

File: src/backend/commands/tablecmds.c [hunk 3]
```diff
@@ -14547,6 +14623,414 @@ ReshuffleRelationData(Relation rel)
...
+		 * Force the use of legacy query optimizer, since PQO will not
+		 * redistribute the tuples if the current and required distributions
+		 * are both RANDOM even when reorganize is set to "true"
+		 */
+		saveOptimizerGucValue = optimizer;
+		optimizer = false;
+
+		if (saveOptimizerGucValue)
+			ereport(LOG,
+					(errmsg("ALTER SET DISTRIBUTED BY: falling back to legacy query optimizer to ensure re-distribution of tuples.")));
```
Comment (danielgustafsson) [inline src/backend/commands/tablecmds.c]: This seems superfluous, if know the process is always falling back to the postgres planner then there is little use logging that.

File: src/include/nodes/parsenodes.h [hunk 4]
```diff
@@ -2107,6 +2114,25 @@ typedef struct PartitionSpec			/* a Partition Specification */
...
+
+typedef struct ExpandStmtSpec
+{
+	NodeTag				type;
+	ExpandMethod		method;
+	Bitmapset			*ps_none;
+	Bitmapset			*ps_root;
+	Bitmapset			*ps_interior;
+	Bitmapset			*ps_leaf;
+	Oid					rootPartOid;
```
Comment (danielgustafsson) [inline src/include/nodes/parsenodes.h]: These members need comments explaining their purposes.

File: gpMgmt/bin/gpexpand [hunk 5]
```diff
@@ -1990,7 +1961,7 @@ UPDATE gp_distribution_policy
         return (' , '.join(name_list), ' , '.join(oid_list))
 
     def perform_expansion(self):
-        """Performs the actual table re-organizations"""
+        """Performs the actual table re-organiations"""
```
Comment (danielgustafsson) [inline gpMgmt/bin/gpexpand]: The previous version was better

File: src/include/nodes/parsenodes.h (hunk 4 above)
Comment (pengzhout) [reply src/include/nodes/parsenodes.h]: added comments.

File: src/backend/commands/tablecmds.c (hunk 3 above)
Comment (pengzhout) [reply src/backend/commands/tablecmds.c]: removed, thanks.

File: gpMgmt/bin/gpexpand (hunk 5 above)
Comment (pengzhout) [reply gpMgmt/bin/gpexpand]: done.

File: src/backend/commands/tablecmds.c (hunk 2 above)
Comment (pengzhout) [reply src/backend/commands/tablecmds.c]: added comments.

File: src/backend/commands/tablecmds.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/commands/tablecmds.c]: @danielgustafsson checked with the team. RESHUFFLE method cannot support randomly distributed partition table, use CTAS for randomly partition table for now.

[the remark this answers is not in the document]
Comment (danielgustafsson) [reply src/backend/commands/tablecmds.c]: What makes a randomly distributed partition table different from a randomly distributed normal table for supporting RESHUFFLE?

[the remark this answers is not in the document]
Comment (pengzhout) [reply src/backend/commands/tablecmds.c]: What I was told is that RESHUFFLE method cannot move data if the target table is randomly distributed and already has the same numsegments with the size of the current cluster.
If the table's numsegments is 2 and size of the current cluster is 4, they will know we need to send tuple to segment 3 and segment 4, so if numsegments already been set to 4, the code inside cannot know the target segment.

for the randomly distributed normal table, its numsegment is expected to be less than cluster size, so RESHUFFLE works, if we re-expand the table again, a "no need to expand" error is reported.

for a randomly distributed partition table, because the PREPARE stage for partition table, all
numsegments has been updated to the size of cluster, so it will trigger RESHUFFLE issue mentioned.

**Comment by pengzhout** [comment]
> I think we could push the 1st part fairly quickly, and we could continue the discussion on ALTER TABLE EXPAND PREPARE and how partitions are handled.

The 1st part has been separated to https://github.com/greenplum-db/gpdb/pull/6301.

**Comment by pengzhout** [comment]
Has seperated the gpexpand part to PR "gpexpand improvement by using newly added features" #6303.

The only thing left in this PR is how to expand a partition by expanding leaf partitions seperately in parallel, there are also voices that it's unnecessary to do such optimization. Let me close this PR
and start a discussion in the dev mail list.
